### PR TITLE
Make test-monitor more robust

### PR DESCRIPTION
**Pull request by yuyichao**
* Keep retry reading from the child if it hasn't exited yet.
* Add a 30 seconds timeout to the command. This helps when the remote side hangs.

Without this the test would hang indefinitely when the gdb remote connection hangs. I initially wrote the non-blocking loop since I thought the child was sleeping waiting for the pipe buffer to clear. I'm not sure that can happen anymore since the EOF should only be sent with the child has closed the file... In any case, I don't think there's a way to implement a timeout with popen easily so I'm just using the non-blocking loop to implement the timeout for now...

### CODE REVIEW

**Comment by Keno** [comment]
We already have an external timeout implemented on CI. For local use, I think I'd rather leave it hanging, so I can go in with gdb to figure out what's going on.

**Comment by yuyichao** [comment]
Would it be OK to control this with an environment variable? Right now this makes it almost impossible and very efficient to run over all the tests. There are enough tests hanging like this that my test run would just get stuck after certain point and I'd have to manually kill the hanging tests to get it unstuck...

**Comment by yuyichao** [comment]
I made the timeout non-default and controlled by an environment variable.

**Comment by DilumAluthge** [comment]
It might be good to document the environment variable somewhere? Maybe in the wiki?

**Comment by yuyichao** [comment]
Added a mention in the usage output. I can also add something to https://github.com/rr-debugger/rr/wiki/Building-And-Installing#timeouts if this is merged...

File: src/test-monitor/test-monitor.cc [hunk 1]
```diff
@@ -25,11 +26,16 @@ static void print_usage(FILE* out) {
...
+        "RUNNING_UNDER_TEST_MONITOR=<pid-of-test-monitor>.\n"
+        "\n"
+        "If the environment variable RR_TEST_DIAGNOSE_TIMEOUT is set to a number\n"
+        "the value will be used as the timeout for each diagnostic commands\n"
+        "(e.g. gdb) after a test failure.\n",
         out);
 }
 
 static unsigned long timeout;
+static unsigned long diagnose_timeout = (unsigned long)-1;
```
Comment (rocallahan) [inline src/test-monitor/test-monitor.cc:38]: `ULONG_MAX`?

File: src/test-monitor/test-monitor.cc [hunk 2]
```diff
@@ -309,6 +353,12 @@ int main(int argc, char* argv[]) {
     fprintf(stderr, "Invalid timeout %s\n", argv[1]);
     return 1;
   }
+  if (char *diagnose_timeout_str = getenv("RR_TEST_DIAGNOSE_TIMEOUT")) {
+      diagnose_timeout = strtoul(diagnose_timeout_str, &endp, 10);
+      if (*endp) {
+          diagnose_timeout = (unsigned long)-1;
```
Comment (rocallahan) [inline src/test-monitor/test-monitor.cc:359]: `ULONG_MAX`?

**Review by rocallahan — CHANGES_REQUESTED** [review]

File: src/test-monitor/test-monitor.cc (hunk 2 above)
Comment (yuyichao) [reply src/test-monitor/test-monitor.cc]: Done. (these are some of the standard macros that I somehow have never got used to using....)

**Pull request merged**